### ISSUE TITLE
Sik-Ho Tang | Review -- BEiT: BERT Pre-Training of Image Transformers.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#79]
Sik-Ho Tang. [Review — BEiT: BERT Pre-Training of Image Transformers](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295).

**Comment by NorbertZheng** [comment]
## Overview
BEiT, Pretraining [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0), Using Masked Image Modeling (MIM).

BEiT: BERT Pre-Training of Image Transformers. BEiT, by Microsoft Research. 2022 ICLR, Over 300 Citations.

Self-Supervised Learning, [BERT](https://sh-tsang.medium.com/59b1684882db), [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1), [Vision Transformer, ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0), [DALL·E](https://sh-tsang.medium.com/review-dall-e-zero-shot-text-to-image-generation-f9de7a383374).

 - Bidirectional Encoder representation from Image Transformers (BEiT) is proposed, where a masked image modeling (MIM) task to pretrain [Vision Transformer](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)s.
 - BEiT first “tokenizes” the original image into visual tokens. Then some image patches are randomly masked and fed into the backbone [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).
 - The pre-training objective is to recover the original visual tokens based on the corrupted image.

**Comment by NorbertZheng** [comment]
## BEiT Architecture
![image](https://user-images.githubusercontent.com/37235023/229813152-c5607174-4922-4fca-a77f-dc3e50a2ffe0.png)
Overview of BEiT pre-training.

### Overall Approach
 - Inspired by [BERT](https://sh-tsang.medium.com/59b1684882db), a pre-training task is proposed, namely, masked image modeling (MIM).
 - MIM uses two views for each images, i.e., image patches, and visual tokens.
 - The image is split into a grid of patches that are the input representation of backbone [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).
 - The image is “tokenized” to discrete visual tokens by the latent codes of discrete VAE, where discrete VAE is from [DALL·E](https://sh-tsang.medium.com/review-dall-e-zero-shot-text-to-image-generation-f9de7a383374).

During pre-training, some proportion of image patches are randomly masked, and fed the corrupted input to [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).
 - The model learns to recover the visual tokens of the original image, instead of the raw pixels of masked patches.

**Comment by NorbertZheng** [comment]
### Image Representation
During pre-training, the images have two views of representations, namely,
 - image patch: input representations,
 - visual tokens output representations.

#### Image Patches
![image](https://user-images.githubusercontent.com/37235023/229815133-47e69876-9172-4a64-9955-ddf27ff78da7.png)
Image Patches (Cut from the first figure).
 - The 2D image of the size $H\times W\times C$ is split into a sequence of patches $x_{p}$ ($p$ is from $1$ to $N$) of the size $P^{2}$, with the number of patch  $N=\frac{HW}{P^{2}}$  patches.
 - The image patches $x_{p}$ are flattened into vectors and are linearly projected which is similar to word embeddings in [BERT](https://sh-tsang.medium.com/59b1684882db).

Particularly, BEiT splits each $224\times 224$ image into a $14\times 14$ grid of image patches, where each patch is $16\times 16$.

#### Visual Tokens
![image](https://user-images.githubusercontent.com/37235023/229816324-c221b358-0d9a-4c3c-83fd-2a0370ae4eef.png)
Visual Tokens (Cut from the first figure).
 - The image is represented as a sequence of discrete tokens obtained by an “image tokenizer”, instead of raw pixels.

Specifically, the image of the size $H\times W\times C$ is tokenized into $z=[z_{1},…,z_{N}]$, where the vocabulary $V=\{1,…,|V|\}$ contains discrete token indices.
 - The image tokenizer learned by discrete variational autoencoder (dVAE), by [DALL·E](https://sh-tsang.medium.com/review-dall-e-zero-shot-text-to-image-generation-f9de7a383374), is directly used.
 - There are two modules during visual token learning, namely, tokenizer and decoder.
 - The tokenizer $q(z|x)$ maps image pixels $x$ into discrete tokens $z$ according to a visual codebook (i.e., vocabulary).
 - The decoder $p(x|z)$ learns to reconstruct the input image $x$ based on the visual tokens $z$.
 - The vocabulary size is set to $|V| = 8192$.

**Comment by NorbertZheng** [comment]
### [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) Backbone
 - Following [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0), the [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) backbone network is used.
 - The input of [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) is a sequence of image patches $x_{i}^{p}$.
 - The patches are then linearly projected to obtain patch embeddings $Ex_{i}^{p}$.
 - The standard learnable 1D position embeddings $E_{pos}$ are added to patch embeddings:
![image](https://user-images.githubusercontent.com/37235023/229818949-1dd08a91-f631-4156-a047-ed74333645b2.png)
 - The encoder contains $L$ layers of [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) blocks:
![image](https://user-images.githubusercontent.com/37235023/229819231-dda6a93b-2573-48e5-9008-b701d3a4d920.png)
 - The output vectors of the last layer is:
![image](https://user-images.githubusercontent.com/37235023/229819441-0707144a-658d-4f20-a522-5f2d44a9e2ca.png)
   which are used as the encoded representations for the image patches, where $h_{i}^{L}$ is the vector of the $i$-th image patch.
 - [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)Base is used, which is a 12-layer [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) with 768 hidden size, and 12 attention heads. The intermediate size of feed-forward networks is 3072.

**Comment by NorbertZheng** [comment]
## BEiT Pretraining: Masked Image Modeling (MIM)
### Masked Image Modeling (MIM)
![image](https://user-images.githubusercontent.com/37235023/229820085-cb879256-f1f7-41bb-bba4-cf1babc1ba88.png)
BEiT Masked Image Modeling (MIM) (Cut from the first figure).
 - After splitting the image into image patches, as described above, approximately 40% image patches are randomly masked, where the masked positions are denoted as $M$. The masked patches are replaced with a learnable embedding $e[M]$. In BEiT, At most 75 patches are masked.
 - Then, the good and masked image patches are input into the L-layer [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).
 - A softmax classifier is used to predict the corresponding visual tokens:
![image](https://user-images.githubusercontent.com/37235023/229820684-98d9bfd6-6b3f-42eb-959a-588b2ef9a049.png)

The pre-training objective is to maximize the log-likelihood of the correct visual tokens $z_{i}$ given the corrupted image:
![image](https://user-images.githubusercontent.com/37235023/229820914-a8219204-2ea4-4b07-963d-b416d7d1782e.png)
 - BEiT is pretrained on the training set of ImageNet-1K.
 - The pre-training runs for about 500k steps (i.e., 800 epochs) with 2k batch size. The 500k training steps take about five days using 16 Nvidia Tesla V100 32GB GPU cards.

**Comment by NorbertZheng** [comment]
### Blockwise Masking
![image](https://user-images.githubusercontent.com/37235023/229821674-97256c73-03d3-4f1a-a442-809f732a1935.png)
Blockwise Masking.
 - Blocks of patches are masked randomly as shown in the figure and algorithm above, instead of ~~masking each patch individually in a random manner (not stable)~~.

**Comment by NorbertZheng** [comment]
### From VAE Perspective
 - The BEiT pre-training can be viewed as variational autoencoder training:
![image](https://user-images.githubusercontent.com/37235023/229822314-1945ab2b-0583-4569-9fdb-e0e0d604f411.png)
 - In the first stage, the image tokenizer is obtained as a discrete variational autoencoder. Specifically, the first stage minimizes the reconstruction loss, with an uniform prior.
 - In the second stage, the prior $p_{\theta}$ is learnt while keeping $q_{\phi}$ and $q_{\Psi}$ fixed.
 - Thus, the above equation is re-written as:
![image](https://user-images.githubusercontent.com/37235023/229822854-7ef5df29-ec29-4720-bdc1-17bf2a088aeb.png)
   where the second term is the proposed BEiT pre-training objective.
